File: 673-number-of-longest-increasing-subsequence/number-of-longest-increasing-subsequence.java
class Solution {
    public int findNumberOfLIS(int[] nums) {
        int n = nums.length;
        int maxlen = 0;
        int[] lis = new int[n];
        int[] count = new int[n];
        int ans = 0;
        Arrays.fill(lis,1);
        Arrays.fill(count,1);

        for(int i=1;i<n;i++){
            for(int j=i-1;j>=0;j--){
                if(nums[i]>nums[j]){
                    if(lis[j]+1 > lis[i]){
                        count[i] = 0;
                        lis[i] = lis[j]+1;
                    }
                    if(lis[j]+1 == lis[i]){
                        count[i] +=count[j];
                    }
                }
            }
        }
        for(int len:lis){
            maxlen = Math.max(maxlen,len);
        }
        for(int a:count){
            System.out.print(a+" ");
        }
        for(int i=0;i<n;i++){
            if(maxlen == lis[i]){
                ans+=count[i];
            }
        }
        return ans;
    }
}

File: 673-number-of-longest-increasing-subsequence/number-of-longest-increasing-subsequence.py
class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        dp = [[1,1] for _ in range(len(nums))]
        n = len(nums)
        hashmap = defaultdict(int)
        for i in range(1,n):
            for j in range(i-1,-1,-1):
                if nums[j]<nums[i]:
                    newlen = 1+dp[j][0]
                    if newlen>dp[i][0]:
                        dp[i] = [newlen,dp[j][1]]
                    elif newlen == dp[i][0]:
                        dp[i][1] += dp[j][1]
        # print(dp)
        for l,v in dp:
            hashmap[l]+=v
        # print(hashmap)
        return hashmap[max(hashmap.keys())]
                
                    